src/Chapter2.hs

```diff
@@ -620,8 +627,14 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
-
+takeEven :: [a] -> [a]
+takeEven n = go 0 [] n
+  where
+    go :: Int -> [a] -> [a] -> [a]
+    go _ acc [] = reverse acc
+    go i acc (x:xs)
+      |mod i 2 == 0 = go (i +1) (x:acc) xs
+      |otherwise = go (i+1) acc xs
```

Comment:
Nice idea on the helper functions :+1:

A few additional thoughts on your idea: as you have only two states of the operation (odd or even), instead of accumulators of the current position number, you can just use `Bool`. And on each next recursion call, you just give it `not acc`. This will remove the necessity to check `mod 2` each time, which would save a couple of functions call on each interaction.

Alternatively, you could use pattern matching in `takeEven` instead. As you need to take every second element, it is possible to represent with a few pattern matching clauses 🙂

Also, calling `reverse` in the end is a bit slow, as `reverse` traverses the whole list. You can avoid calling `reverse` by adding new elements directly to the result:

```haskell
| mod i 2 == 0 = x : go (i +1) acc xs
```

After that refactoring you can notice that the `acc` variable becomes redundant :recycle: